suite-native/discovery/src/discoveryThunks.ts

```diff
@@ -1,24 +1,247 @@
 import { createThunk } from '@suite-common/redux-utils';
 import { TrezorDevice } from '@suite-common/suite-types';
 import {
+    accountsActions,
     createDiscoveryThunk,
     DISCOVERY_MODULE_PREFIX,
-    startDiscoveryThunk,
+    filterUnavailableNetworks,
+    MAX_ACCOUNTS_PER_NETWORK,
+    selectAccountByNetworkAndPath,
+    selectDiscoveryForDevice,
+    updateDiscovery,
 } from '@suite-common/wallet-core';
+import TrezorConnect from '@trezor/connect';
+import { DiscoveryItem } from '@suite-common/wallet-types';
+import { getDerivationType } from '@suite-common/wallet-utils';
+import { Network } from '@suite-common/wallet-config';
+import { DiscoveryStatus } from '@suite-common/wallet-constants';
 
-export const createAndStartDiscoveryThunk = createThunk(
-    `${DISCOVERY_MODULE_PREFIX}/createAndStart`,
+import { DeviceAccessMutex, getBundleDescriptors } from './utils';
+
+const DISCOVERY_ROUND_SIZE = 3;
+
+type DiscoveryDescriptorItem = DiscoveryItem & { descriptor: string };
+
+const finishNetworkTypeDiscoveryThunk = createThunk(
+    `${DISCOVERY_MODULE_PREFIX}/incrementDiscoveryProgressThunk`,
+    (_, { dispatch, getState }) => {
+        const discovery = selectDiscoveryForDevice(getState());
+
+        if (!discovery) {
+            return;
+        }
+
+        const finishedNetworksCount = discovery.loaded + 1;
+        dispatch({
+            type: updateDiscovery.type,
+            payload: {
+                ...discovery,
+                loaded: finishedNetworksCount,
+            },
+        });
+
+        if (finishedNetworksCount === discovery.networks.length) {
+            dispatch({
+                type: updateDiscovery.type,
+                payload: {
+                    ...discovery,
+                    status: DiscoveryStatus.COMPLETED,
+                },
+            });
+        }
+    },
+);
+
+const discoverAccountsByDescriptorThunk = createThunk(
+    `${DISCOVERY_MODULE_PREFIX}/discoverAccountsByXpubThunk`,
     async (
-        { deviceState, device }: { deviceState: string; device: TrezorDevice },
+        {
+            descriptorsBundle,
+            deviceState,
+        }: {
+            descriptorsBundle: DiscoveryDescriptorItem[];
+            deviceState: string;
+        },
         { dispatch },
     ) => {
+        let isFinalRound = false;
+
+        if (!descriptorsBundle) {
+            isFinalRound = true;
+        }
+
+        // eslint-disable-next-line no-restricted-syntax
+        for (const bundleItem of descriptorsBundle) {
+            // eslint-disable-next-line no-await-in-loop
```

Comment:
This is actually legit rule, you should probably use `Promise.all()` if possible. Since we don't care about the order of the accounts too much (or do we?) this will further optimize the speed of discovery

Reply:
This loop has to be sequential since the previous round always determines if the following one is executed. So There is need for `for of `loop.